### PR TITLE
Improve switching between rich and plain editing modes

### PR DESCRIPTION
**Pull request by alunturner**
Changes add:

- the ability to switch back and forth between rich and plain text modes whilst maintaining formatting (nb there is a known issue with formatting tags containing white space inside tags and how they are converted to markdown, raised here: [issue in rich text editor repo](https://github.com/matrix-org/matrix-rich-text-editor/issues/407)). Example conversions shown below
Starting in rich text mode, enter some formatted text, click convert
Element___Alun_Turner
Plain text contains the for
Element___Alun_Turner
matted text converted to markdown, then click convert to go back to rich text mode
Back to the original version
Element___Alun_Turner

- updates rich text composer dependency to 0.12.0 (@matrix-org/matrix-wysiwyg)
- use of conversion functions from that version to convert between rich and plain text, these are imported dynamically to avoid this code being imported when the rich text labs feature is disabled
- improves handling of enter keypresses when `ctrlEnterTosend` setting is true in plain text editor
- changes the message event content when using the new editor to use output from the new conversion functions
- adds tests for the changes to the plain text editor

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-react-sdk/blob/develop/CONTRIBUTING.md))

---
Here's what your changelog entry will look like:

## ✨ Features
 * Improve switching between rich and plain editing modes ([\#9776](https://github.com/matrix-org/matrix-react-sdk/pull/9776)). Contributed by @alunturner.

### CODE REVIEW

File: src/components/views/rooms/MessageComposer.tsx [hunk 1]
```diff
@@ -15,6 +15,7 @@ limitations under the License.
 */
 
 import React, { createRef, ReactNode } from "react";
+import { richToPlain, plainToRich } from "@matrix-org/matrix-wysiwyg";
```
Comment (t3chguy) [inline src/components/views/rooms/MessageComposer.tsx:18]: Is this import safe to use here without the dependency on https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry

Comment (alunturner) [reply src/components/views/rooms/MessageComposer.tsx:18]: No you're right this is going to cause the same issue that was looked at this morning. I'm sorting out lint/TS errors then will sort this.

File: src/components/views/rooms/MessageComposer.tsx [hunk 2]
```diff
@@ -333,8 +332,12 @@ export class MessageComposer extends React.Component<IProps, IState> {
         this.messageComposerInput.current?.sendMessage();
 
         if (this.state.isWysiwygLabEnabled) {
+            // due to this function using wysiwyg functions targetting ES6,
+            // we need to import dynamically
+            const { sendMessage } = await import("./wysiwyg_composer/");
```
Comment (florianduros) [inline src/components/views/rooms/MessageComposer.tsx:337]: FYI, we don't need to have a dynamic import for `sendMessage`, it doesn't have dependency with the wysiwyg lib

Comment (florianduros) [inline src/components/views/rooms/MessageComposer.tsx:337]: I think that will be a good idea to wrap this dynamic import in the RTE files.

You can take a look at the `SendWysiwygComposer` composer for example

File: src/components/views/rooms/MessageComposer.tsx [hunk 3]
```diff
@@ -359,14 +362,21 @@ export class MessageComposer extends React.Component<IProps, IState> {
...
-        this.setState((state) => ({
-            isRichTextEnabled: !state.isRichTextEnabled,
-            initialComposerContent: !state.isRichTextEnabled
-                ? state.composerContent
-                : // TODO when available use rust model plain text
-                  htmlToPlainText(state.composerContent),
-        }));
+    private onRichTextToggle = async () => {
+        // due to this repo targetting ES6, we need to import dynamically
+        const conversionFunctions = await import("@matrix-org/matrix-wysiwyg");
```
Comment (florianduros) [inline src/components/views/rooms/MessageComposer.tsx:367]: I think that will be a good idea to wrap this dynamic import in the RTE files.

You can take a look at the SendWysiwygComposer composer for example

**Review by florianduros — CHANGES_REQUESTED** [review]
Looks great, minor changes requested :)

**Comment by t3chguy** [comment]
It seems to convert strikethrough to `~~TEXT~~` and it works for sending but if the user enters that manually then no such luck

**Comment by alunturner** [comment]
> It seems to convert strikethrough to `~~TEXT~~` and it works for sending but if the user enters that manually then no such luck

Can you confirm that you were typing that text into the plain text editor (ie when the formatting buttons are not showing above the composer box)? When I type that text (in between the backticks) into the plain editor and send it then it is displays as struck through when sent.

From my understanding this is the desired behaviour, but it does raise a good point about what should happen if the user types valid markdown into the rich text editor.

**Review by florianduros — APPROVED** [review]

**Review by andybalaam — APPROVED** [review]
Approved, thanks!

**Pull request merged**